File: python/change/change.py
def find_fewest_coins(coins, target):
    """
    Find the fewest number of coins needed to reach a target value using the given coin denominations.

    Args:
        coins (list of int): A list of coin denominations available for making change.
        target (int): The target value to reach using the coins.

    Returns:
        list of int: A sorted list of coin denominations that make up the fewest coins to reach the target.

    Raises:
        ValueError: If the target value is negative or cannot be reached with the given coins.

    The function uses dynamic programming to determine the optimal combination of coins that
    minimizes the total number of coins needed to reach the target value. It returns the list
    of coin denominations in ascending order.

    If it's not possible to reach the target value using the given coins, a ValueError is raised.
    """
    # Check if the target is negative, and raise a ValueError if it is.
    if target < 0:
        raise ValueError("target can't be negative")

    # Initialize a dynamic programming (DP) list to store the optimal coin combinations.
    dp = [[] for _ in range(target + 1)]

    # Loop through all possible values from 0 to (target - 1).
    for i in range(target):
        # If i is 0 or dp[i] is not an empty list, continue.
        if i == 0 or dp[i] != []:
            # Iterate through the available coin denominations.
            for coin in coins:
                # Check if adding the current coin to i doesn't exceed the target
                # and if adding this coin results in a shorter list of coins.
                if i + coin <= target and (dp[i + coin] == [] or len(dp[i + coin]) > len(dp[i]) + 1):
                    # Update the DP list with the new, shorter combination of coins.
                    dp[i + coin] = dp[i] + [coin]

    # If the target is greater than 0 and dp[target] is still an empty list,
    # it means the target cannot be reached with the given coin denominations.
    if target > 0 and dp[target] == []:
        raise ValueError("can't make target with given coins")

    # Return the sorted list of coins that make up the fewest coins for the target.
    return sorted(dp[target])

File: javascript/change/change.js
export class Change {
  calculate(coinArray, target) {
    // If target is 0, no change is needed
    if (target === 0) {
      return [];
    }

    // Negative totals are not allowed
    if (target < 0) {
      throw new Error("Negative totals are not allowed.");
    }

    // Ensure the target can be represented with the given coins
    if (target < Math.min(...coinArray)) {
      throw new Error(
        "The total " + target + " cannot be represented in the given currency.",
      );
    }

    // Initialize an array to store minimum coins needed for each target value
    const dp = new Array(target + 1).fill(Infinity);
    dp[0] = 0; // Zero coins needed for target value of 0

    // Compute minimum coins needed for each target value up to 'target'
    for (const coin of coinArray) {
      for (let i = coin; i <= target; i++) {
        dp[i] = Math.min(dp[i], dp[i - coin] + 1);
      }
    }

    // If no valid solution exists for the target, throw an error
    if (dp[target] === Infinity) {
      throw new Error(
        "The total " + target + " cannot be represented in the given currency.",
      );
    }

    // Find the coins used for the minimum number of coins
    const coinsUsed = [];
    let remaining = target;

    while (remaining > 0) {
      for (const coin of coinArray) {
        if (remaining >= coin && dp[remaining - coin] + 1 === dp[remaining]) {
          coinsUsed.push(coin);
          remaining -= coin;
          break;
        }
      }
    }

    return coinsUsed;
  }
}
